include/deal.II/lac/solver_cg.h

```diff
@@ -389,6 +462,481 @@ SolverCG<VectorType>::compute_eigs_and_cond(
+    template <typename VectorType,
+              typename MatrixType,
+              typename PreconditionerType,
+              typename = int>
+    struct IterationWorker
+      : public IterationWorkerBase<VectorType, MatrixType, PreconditionerType>
+    {
+      IterationWorker(const MatrixType &        A,
+                      const PreconditionerType &preconditioner,
+                      const bool                flexible,
+                      VectorMemory<VectorType> &memory,
+                      VectorType &              x)
+        : IterationWorkerBase<VectorType, MatrixType, PreconditionerType>(
+            A,
+            preconditioner,
+            flexible,
+            memory,
+            x)
+      {}
+    };
+
+
+
+    // Internal function to run one iteration of the conjugate gradient solver
+    // for matrices and preconditioners that support interleaving the vector
+    // updates with the matrix-vector product.
+    template <typename VectorType,
+              typename MatrixType,
+              typename PreconditionerType>
+    struct IterationWorker<
+      VectorType,
+      MatrixType,
+      PreconditionerType,
+      typename std::enable_if<
+        supports_vmult_with_std_functions<MatrixType,
+                                          VectorType,
+                                          PreconditionerType>::value,
+        int>::type>
+      : public IterationWorkerBase<VectorType, MatrixType, PreconditionerType>
+    {
+      static const unsigned int grain_size = 128;
+
+      IterationWorker(const MatrixType &        A,
+                      const PreconditionerType &preconditioner,
+                      const bool                flexible,
+                      VectorMemory<VectorType> &memory,
+                      VectorType &              x)
+        : IterationWorkerBase<VectorType, MatrixType, PreconditionerType>(
+            A,
+            preconditioner,
+            flexible,
+            memory,
+            x)
+      {}
+
+      void
+      do_iteration(const unsigned int iteration_index)
+      {
+        using Number = typename VectorType::value_type;
+
+        const auto operation_before_loop = [&](const unsigned int start_range,
+                                               const unsigned int end_range) {
+          Number *                       x = this->x.begin() + start_range;
+          Number *                       r = this->r.begin() + start_range;
+          Number *                       p = this->p.begin() + start_range;
+          Number *                       v = this->v.begin() + start_range;
+          std::array<Number, grain_size> prec_r;
+          if (iteration_index == 1)
+            {
+              for (unsigned int j = start_range; j < end_range; j += grain_size)
+                {
+                  const unsigned int length =
+                    std::min(grain_size, end_range - j);
+                  this->preconditioner.apply_to_subrange(j,
```

Comment:
Are you planing to add a specialization also for the non-preconditioned case?

Reply:
Maybe, it depends on the time I have available. One could imagine a specialization for `PreconditionIdentity` and `DiagonalMatrix` that avoids stepping into those functions at all and merely does the update.